saffron/src/query.rs

```diff
@@ -0,0 +1,292 @@
+use std::marker::PhantomData;
+
+use crate::utils::decode_from_field_elements;
+use ark_ff::PrimeField;
+use ark_poly::{EvaluationDomain, Evaluations, Radix2EvaluationDomain};
+use rayon::prelude::*;
+use thiserror::Error;
+
+#[derive(Clone, Debug)]
+/// Represents the bytes a user query
+pub struct QueryBytes {
+    pub start: usize,
+    pub len: usize,
+}
+
+#[derive(Copy, Clone, PartialEq, PartialOrd, Eq, Ord, Debug)]
+/// We store the data in a vector of vector of field element
+/// The inner vector represent polynomials
+pub struct FieldElt {
+    /// the index of the polynomial the data point is attached too
+    pub poly_index: usize,
+    /// the index of the root of unity the data point is attached too
+    pub eval_index: usize,
+    domain_size: usize,
+    n_polys: usize,
+}
+/// Represents a query in term of Field element
+#[derive(Debug)]
+pub struct QueryField<F> {
+    pub start: FieldElt,
+    /// how many bytes we need to trim from the first chunk
+    /// we get from the first field element we decode
+    pub leftover_start: usize,
+    pub end: FieldElt,
+    /// how many bytes we need to trim from the last chunk
+    /// we get from the last field element we decode
+    pub leftover_end: usize,
+    tag: PhantomData<F>,
+}
+
+pub struct IndexQuery {
+    pub chunks: Vec<Vec<usize>>,
+}
+
+pub struct QueryResult<F> {
+    pub chunks: Vec<Vec<(usize, F)>>,
+}
+
+impl<F: PrimeField> QueryResult<F> {
+    pub fn as_evaluations(
+        &self,
+        domain: Radix2EvaluationDomain<F>,
+    ) -> Vec<Evaluations<F, Radix2EvaluationDomain<F>>> {
+        self.chunks
+            .par_iter()
+            .map(|chunk| {
+                let mut evals = vec![F::zero(); domain.size()];
+                chunk.iter().for_each(|(j, val)| {
+                    evals[*j] = *val;
+                });
+                Evaluations::from_vec_and_domain(evals, domain)
+            })
+            .collect()
+    }
+}
+
+impl<F: PrimeField> QueryField<F> {
+    fn n_polys(&self) -> usize {
+        self.start.n_polys
+    }
+
+    pub fn apply(&self, data: &[Vec<F>]) -> QueryResult<F> {
+        let mut chunks = vec![Vec::new(); self.n_polys()];
+        self.start
+            .into_iter()
+            .take_while(|x| x <= &self.end)
+            .for_each(|x| {
+                let value = data[x.poly_index][x.eval_index];
+                chunks[x.poly_index].push((x.eval_index, value));
+            });
+        QueryResult { chunks }
+    }
+
+    pub fn result_decoder(self) -> impl Fn(&QueryResult<F>) -> Vec<u8> {
```

Comment:
Why this signature?

Reply:
See https://github.com/o1-labs/proof-systems/pull/3017

Reply:
@dannywillems Because all I ever wanted since I was a child was to become a functional programmer :upside_down_face: